src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+        """
+        Sets defocus.
+
+        :param df: Defocus value in the direction perpendicular to df1.
+        """
+
+        self.defocus2 = df
+
+    def set_angle(self, angle):
+        """
+        Sets angle.
+
+        :param angle: Angle (in degrees) between df1 and the x-axis.
+        """
+
+        self.angle = angle
+
+    def generate_ctf(self):
+        """
+        Generates internal representation of the Contrast Transfer Function using parameters from this instance.
+        """
+
+        astigmatism_angle = np.reshape(
+            np.repeat(
+                self.angle, np.multiply(self.theta.shape[0], self.theta.shape[1])
+            ),
+            self.theta.shape,
+        )
+        defocus_sum = np.reshape(
+            np.repeat(
+                self.defocus1 + self.defocus2,
+                np.multiply(self.theta.shape[0], self.theta.shape[1]),
+            ),
+            self.theta.shape,
+        )
+        defocus = defocus_sum + np.multiply(
+            self.defocus1 - self.defocus2, np.cos(2 * (self.theta - astigmatism_angle))
+        )
+        defocus_factor = np.pi * self.lmbd * np.multiply(self.r_ctf, defocus) / 2
+        amplitude_contrast_term = np.divide(
+            self.amplitude_contrast, np.sqrt(1 - np.square(self.amplitude_contrast))
+        )
+        chi = (
+            defocus_factor
+            - np.pi
+            * np.power(self.lmbd, 3)
+            * self.cs
+            * np.power(10, 6)
+            * np.square(self.r_ctf)
+            / 2
+            + amplitude_contrast_term
+        )
+        h = -np.sin(chi)
+        self.h = h
+
+    def ctf_preprocess(self, micrograph, block_size):
+        """
+        Preprocess CTF of micrograph using block_size.
+
+        :param micrograph: Micrograph as numpy array. #NOTE looks like F order
+        :param blocksize: Size of the square blocks to partition micrograph.
+        :return: Numpy array of blocks extracted from the micrograph.
+        """
+
+        # verify block_size is even
+        block_size = block_size - (block_size % 2)
+
+        size_x = micrograph.shape[1]
+        size_y = micrograph.shape[0]
+
+        step_size = block_size // 2
+        range_y = size_y // step_size - 1
+        range_x = size_x // step_size - 1
+
+        block_list = [
+            micrograph[
+                i * step_size : (i + 2) * step_size, j * step_size : (j + 2) * step_size
+            ]
+            for j in range(range_y)
+            for i in range(range_x)
+        ]
+        block = np.asarray(block_list, dtype=micrograph.dtype)
+
+        block_sum = np.sum(np.sum(block, axis=-1), axis=-1)
+        block_sum = np.reshape(block_sum, (block.shape[0], 1, 1))
+        block_sum = np.tile(block_sum, (1, block_size, block_size))
+        block = block - (
+            block_sum / (block_size ** 2)
+        )  # equals to the matlab version (11-7-19)
+
+        return block
+
+    def ctf_tapers(self, N, R, L):
```

Comment:
Cool, thanks for suggesting.  I think that should be better too.

I was able to replace this after some experimentation.  Once I found an incantation to get resolution for scipy, ASPIRE's implementation turned out to be the negated transpose of scipy... :joy: :tada:

Reply:
Huh… Transpose I get (this part of SciPy, for some reason, seems to adopt column-major ordering) but sign shouldn't matter. Tests should still pass?

Reply:
It will not matter in the current code (eventually squared), but I wasn't sure if there are other components that would care about the convention in the future (plus it made me laugh). I'll read this is unlikely and take out the negative I put in to match the original implementation.

If you want me to cut `tapers` and just call scipy from `estimate_ctf` we can, just lmk.  Personally I think it might be useful to have that layer in there in case someone wants to play/change it, or at least keep the code comments out of a more main-like method.

Reply:
No I think this is good.